substrate/client/network/types/src/ed25519.rs

```diff
@@ -0,0 +1,551 @@
+		Self::try_from_bytes(&mut kp.encode())
+			.expect("ed25519_dalek in substrate & litep2p to use the same format")
+	}
+}
+
+impl From<Keypair> for litep2p_ed25519::Keypair {
+	fn from(kp: Keypair) -> Self {
+		Self::decode(&mut kp.to_bytes())
+			.expect("ed25519_dalek in substrate & litep2p to use the same format")
+	}
+}
+
+impl From<libp2p_ed25519::Keypair> for Keypair {
+	fn from(kp: libp2p_ed25519::Keypair) -> Self {
+		Self::try_from_bytes(&mut kp.to_bytes())
+			.expect("ed25519_dalek in substrate & libp2p to use the same format")
+	}
+}
+
+impl From<Keypair> for libp2p_ed25519::Keypair {
+	fn from(kp: Keypair) -> Self {
+		Self::try_from_bytes(&mut kp.to_bytes())
+			.expect("ed25519_dalek in substrate & libp2p to use the same format")
+	}
+}
+
+/// Demote an Ed25519 keypair to a secret key.
+impl From<Keypair> for SecretKey {
+	fn from(kp: Keypair) -> SecretKey {
+		SecretKey(kp.0.to_bytes())
+	}
+}
+
+/// Promote an Ed25519 secret key into a keypair.
+impl From<SecretKey> for Keypair {
+	fn from(sk: SecretKey) -> Keypair {
+		let signing = ed25519::SigningKey::from_bytes(&sk.0);
+		Keypair(signing)
+	}
+}
+
+/// An Ed25519 public key.
+#[derive(Eq, Clone)]
+pub struct PublicKey(ed25519::VerifyingKey);
+
+impl fmt::Debug for PublicKey {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		f.write_str("PublicKey(compressed): ")?;
+		for byte in self.0.as_bytes() {
+			write!(f, "{byte:x}")?;
+		}
+		Ok(())
+	}
+}
+
+impl cmp::PartialEq for PublicKey {
+	fn eq(&self, other: &Self) -> bool {
+		self.0.as_bytes().eq(other.0.as_bytes())
+	}
+}
+
+impl hash::Hash for PublicKey {
+	fn hash<H: hash::Hasher>(&self, state: &mut H) {
+		self.0.as_bytes().hash(state);
+	}
+}
+
+impl cmp::PartialOrd for PublicKey {
+	fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl cmp::Ord for PublicKey {
+	fn cmp(&self, other: &Self) -> cmp::Ordering {
+		self.0.as_bytes().cmp(other.0.as_bytes())
+	}
+}
+
+impl PublicKey {
+	/// Verify the Ed25519 signature on a message using the public key.
+	pub fn verify(&self, msg: &[u8], sig: &[u8]) -> bool {
+		ed25519::Signature::try_from(sig).and_then(|s| self.0.verify(msg, &s)).is_ok()
+	}
+
+	/// Convert the public key to a byte array in compressed form, i.e.
+	/// where one coordinate is represented by a single bit.
+	pub fn to_bytes(&self) -> [u8; 32] {
+		self.0.to_bytes()
+	}
+
+	/// Try to parse a public key from a byte array containing the actual key as produced by
+	/// `to_bytes`.
+	pub fn try_from_bytes(k: &[u8]) -> Result<PublicKey, DecodingError> {
+		let k =
+			<[u8; 32]>::try_from(k).map_err(|e| DecodingError::PublicKeyParseError(Box::new(e)))?;
+		ed25519::VerifyingKey::from_bytes(&k)
+			.map_err(|e| DecodingError::PublicKeyParseError(Box::new(e)))
+			.map(PublicKey)
+	}
+
+	/// Convert public key to `PeerId`.
+	pub fn to_peer_id(&self) -> PeerId {
+		litep2p::PeerId::from(litep2p::crypto::PublicKey::Ed25519(self.clone().into())).into()
```

Comment:
Can we not directly construct the PeerId?

Reply:
This would require hashing `PublicKey` type serialized as protobuf, and I tried to avoid including the protobuf schema into the already bloated wrapper sources.